Tickets/dm 51658

dsanmartim opened the pull request.

dsanmartim committed:
Add a ``daytime_checkout/lsstcam_checkout.py`` script and unit test.


Fix punctiation for lsstcam_checkout.py script.

dsanmartim committed:
Add a ``daytime_checkout/lsstcam_checkout.py`` script and unit test.

dsanmartim committed:
Add news fragment.

dsanmartim committed:
Add news fragment.

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,584 @@
+            try:
+                mtptg_state = await self.mtcs.get_state("mtptg")
+            except Exception as e:
+                raise RuntimeError(f"Failed to retrieve MTPtg state: {e}") from e
```
please, add punctuation at the end of the phrase.

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,584 @@
+            self.available_filters = await self.lsstcam.get_available_filters()
+            self.log.info(f"Available filters: {self.available_filters}")
+        except Exception as e:
+            self.log.warning(f"Could not retrieve available filters: {e}")
```
please, add punctuation at the end of the phrase.

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,584 @@
+                self.current_filter = inst_filter
+        except Exception:
+            inst_filter = "UNKNOWN"
+        self.log.info(f"Engineering test completed with filter {inst_filter}")
```
please, add punctuation at the end of the phrase.

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,584 @@
+        ingestion_events = []
+        start_time = utils.current_tai()
+
+        self.log.info(f"Waiting for OODS ingestion events for obsid {expected_obsid}")
```
please, add punctuation at the end of the phrase.

tribeiro reviewed (commented):
I have some comments I hope you will consider...


One general comment, I noticed several (but not all) of your log messages are missing punctuation. Please, make sure you add a "." at the end of the phrases.

dsanmartim commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,584 @@
+
+        self.log.info(f"Waiting for OODS ingestion events for obsid {expected_obsid}")
+
+        while utils.current_tai() - start_time < self.ingestion_timeout:
```
I added the asynccontextmanager and simplified things a little, but it still has some complexity to check for all the possible raft/sensor combinations. Not sure if that is what you were expecting.

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,608 @@
+        try:
+            # Execute the image acquisition inside the context block
+            yield
+            success = True
```
instead of doing this, you could do:

```
try:
    yield
except Exception:
    raise
else:
    # do stuff here.
```

The `else` block only executes if the `try` was successful.

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,608 @@
+            if not success:
+                return
+
+            # Collect all fresh ingestion events for the latest exposure.
```
I would suggest splitting all these activities into their own methods so the operation here is cleaner and easier to read...

dsanmartim committed:
Add a ``daytime_checkout/lsstcam_fes_exercise.py`` script and unit test.

dsanmartim committed:
Add a ``daytime_checkout/lsstcam_fes_exercise.py`` script and unit test.

dsanmartim committed:
fixup! Add a ``daytime_checkout/lsstcam_checkout.py`` script and unit test.

dsanmartim committed:
Simplify LSSTCam FES readiness checks

Replace the ignore-list handling and broad all-enabled checks with direct state checks for the minimum components needed by the filter exercise: MTCamera, MTPtg, and MTRotator must be enabled, while MTMount can be either enabled or disabled.

edennihy reviewed (approved):
lsstcam_checkout script looks good!

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,607 @@
+INGESTION_TIMEOUT = 15  # max time to wait for all raft/sensors ingestion events
+
+# Expected ingestion counts by image type
+EXPECTED_BIAS_INGEST_SCIENCE = 189
```
You define these here and later just uses them to set the class attributes. I think you can skip this and just define the values directly in the class....

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,607 @@
+
+    def set_metadata(self, metadata):
+        """Set estimated duration and metadata."""
+        read_out_time = getattr(self.lsstcam, "read_out_time", 2.0)
```
Why use `getattr` here? You should be able to just do `read_out_time = self.lsstcam.read_out_time`.

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,607 @@
+    def set_metadata(self, metadata):
+        """Set estimated duration and metadata."""
+        read_out_time = getattr(self.lsstcam, "read_out_time", 2.0)
+        shutter_time = getattr(self.lsstcam, "shutter_time", 1.0)
```
Same as above, why use `getattr`?

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,607 @@
+        """Set estimated duration and metadata."""
+        read_out_time = getattr(self.lsstcam, "read_out_time", 2.0)
+        shutter_time = getattr(self.lsstcam, "shutter_time", 1.0)
+        per_exposure_time = self.ingestion_timeout + read_out_time + shutter_time
```
You only use `read_out_time` and `shutter_time` once here.. do you really need to create these variables?

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,607 @@
+
+        metadata.duration = 2 * per_exposure_time
+        metadata.instrument = "LSSTCam"
+        metadata.filter = self.current_filter
```
`self.current_filter` is `None` by default... I don't think `metadata.filter` supports `None` you might want to check and only set if it it not `None`... Also, will it be set when this is called? I don't think it will, you might want to skip this altogether...

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,607 @@
+        metadata.duration = 2 * per_exposure_time
+        metadata.instrument = "LSSTCam"
+        metadata.filter = self.current_filter
+        metadata.survey = self.program
```
Again, make sure `self.program` is not `None` before setting it here.

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,607 @@
+
+        Raises
+        ------
+        AssertionError
```
You are not the one doing the raise so, in principle, you don't need to add it here...

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,607 @@
+            self.current_filter = await self.lsstcam.get_current_filter()
+            self.log.info(f"Current filter in beam: {self.current_filter}.")
+        except Exception as e:
+            self.log.warning(f"Could not retrieve current filter: {e}.")
```
Consider doing this:

```
try:
    ...
except Exception:
    self.log.warning(..., exc_info=True)
```

This will add the full traceback to the message.

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,607 @@
+            self.available_filters = await self.lsstcam.get_available_filters()
+            self.log.info(f"Available filters: {self.available_filters}.")
+        except Exception as e:
+            self.log.warning(f"Could not retrieve available filters: {e}.")
```
same as above, use `exc_info=True` to log the full exception.

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,607 @@
+        try:
+            inst_filter = await self.lsstcam.get_current_filter()
+        except Exception as e:
+            self.log.warning(
```
again, use `exc_info=True`

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_fes_exercise.py:
```diff
@@ -0,0 +1,381 @@
+        change_timeout = getattr(self.lsstcam, "filter_change_timeout", 120)
+        metadata.duration = 2 * float(change_timeout) + SLEEP_BETWEEN_FILTER_CHANGES
+        metadata.instrument = "LSSTCam"
+        metadata.filter = self.final_filter
```
Same as in the other script, I don't think this can be `None` so make sure you check before you set...  might be better to just leave it alone...

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_checkout.py:
```diff
@@ -0,0 +1,607 @@
+        science_pairs = set()
+        guider_pairs = set()
+        wfs_pairs = set()
+        science_pattern = re.compile(r"^S\d{2}$")
```
You might want to make these class attributes or constants instead of creating them at every call as these tend to be expensive operations.

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_fes_exercise.py:
```diff
@@ -0,0 +1,381 @@
+                f"Requested final filter {self.final_filter} is not currently available."
+            )
+
+        final_filter = self.final_filter.strip()
```
If you need to strip the `final_filter` don't you need to do that earlier? line in line 251 above?

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_fes_exercise.py:
```diff
@@ -0,0 +1,381 @@
+    "y": "y_10",
+}
+
+SLEEP_BETWEEN_FILTER_CHANGES = 120  # seconds
```
Consider renaming it to something like `DELAY_BETWEEN_FILTER_CHANGES` or `INTERVAL_BETWEEN_FILTER_CHANGES`...

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_fes_exercise.py:
```diff
@@ -0,0 +1,381 @@
+
+        await self._assert_state(self.lsstcam, "mtcamera", salobj.State.ENABLED)
+
+    async def _assert_state(
```
This is called `assert` but it raising a `RuntimeError` consider renaming it or raising `AssertionError`.

But actually, it is a bit weird that you have to pass the `remote_group`, that is already part of the class to the method. I think you can simplify this a bit by passing the state instead. Something like:

```
def _check_component_state(self, component, current_state, required_state):
    if current_state != required_state:
        raise RuntimeError(
            f"{component} must be {required_state.name} before exercising filters. "
            f"Current state is {current_state.name}."
        )
```

Then you would call it like:

```
component = "mtptg"
current_state = await self.mtcs.get_state(component)
self._check_component_state(component, current_state, required_state)
```

You could even consider making `required_state` a set so you can use this later for the MTMount case.

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_fes_exercise.py:
```diff
@@ -0,0 +1,381 @@
+            self.current_filter = target_filter
+        except Exception as exc:
+            raise RuntimeError(
+                f"Failed to change filter to {target_filter}: {exc}."
```
you might want to remove the `exc` from the exception string since you are already raising it from the exception.

tribeiro commented on python/lsst/ts/maintel/standardscripts/daytime_checkout/lsstcam_fes_exercise.py:
```diff
@@ -0,0 +1,381 @@
+            return []
+
+        if isinstance(available_filters, str):
+            return [flt.strip() for flt in available_filters.split(",") if flt.strip()]
```
consider just using `f` instead of `flt`...

tribeiro reviewed (changes requested):
I have some comments I hope you will consider before merging. 

One general comment, I noticed some log messages are missing punctuation while others do have punctuation. Please, make sure to punctuate all the log messages.

dsanmartim committed:
Adrress PR comments for lsstcam_checkout.py.

dsanmartim committed:
Adrress PR comments for lsstcam_fes_exercise.py.

dsanmartim committed:
Make LSSTCam checkout TCS sync behavior explicit

Configure LSSTCam with tcs_ready_to_take_data=None for the daytime checkout and document that the engineering frame is intentionally taken without TCS synchronization.

dsanmartim committed:
Replace LSSTCam checkout ENGTEST with DARK frames.
